pkg/util/tracing/collector/collector.go

```diff
@@ -53,59 +50,125 @@ func New(
 	}
 }
 
-// GetSpanRecordingsFromCluster returns the inflight span recordings from all
-// nodes in the cluster.
-// This method does not distinguish between requests for local and remote
-// inflight spans, and relies on gRPC short circuiting local requests.
-func (t *TraceCollector) GetSpanRecordingsFromCluster(
-	ctx context.Context, traceID uint64,
-) ([]tracingpb.RecordedSpan, error) {
-	var res []tracingpb.RecordedSpan
-	ns, err := nodesFromNodeLiveness(ctx, t.nodeliveness)
-	if err != nil {
-		return res, err
+// TraceCollectorIterator can be used to return RecordedSpans from all live
+// nodes in the cluster, in a streaming manner. The iterator buffers the
+// RecordedSpans of one node at a time.
+type TraceCollectorIterator struct {
+	collector *TraceCollector
+
+	ctx context.Context
+
+	traceID uint64
+
+	// liveNodes represents all the nodes in the cluster that are considered live,
+	// and will be contacted for inflight trace spans by the iterator.
+	liveNodes []roachpb.NodeID
+
+	// curNodeIndex maintains the node from which the iterator has pulled inflight
+	// span recordings and buffered them in `recordedSpans` for consumption via
+	// the iterator.
+	curNodeIndex int
+
+	// recordedSpanIndex maintains the current position of of the iterator in the
+	// list of recorded spans. The recorded spans that the iterator points to are
+	// buffered in `recordedSpans`.
+	recordedSpanIndex int
+
+	// recordedSpans represents all recorded spans for a given node currently
+	// accessed by the iterator.
+	recordedSpans []tracingpb.RecordedSpan
+
+	iterErr error
+}
+
+// Start fetches the live nodes in the cluster, and configures the underlying
+// iterator that is used to access recorded spans in a streaming fashion.
+func (t *TraceCollector) Start(ctx context.Context, traceID uint64) *TraceCollectorIterator {
+	tc := &TraceCollectorIterator{ctx: ctx, traceID: traceID, collector: t, curNodeIndex: -1}
+	tc.liveNodes, tc.iterErr = nodesFromNodeLiveness(ctx, t.nodeliveness)
+	return tc
+}
+
+// Valid returns whether the iterator is in a valid state, and can return a
+// RecordedSpan.
+// If we have exhausted the buffered recordedSpans, the method will poll other
+// live nodes for inflight spans.
+func (i *TraceCollectorIterator) Valid() bool {
```

Comment:
So typically the contract for iterators would only "do the work" as part of .Next(), or .Start(). See how the various iterators in CRDB behave, .Start() would be similar to .Seek() in pebble in that it "positions" the iterator in a valid state, if possible. 

I see that calling .Valid() repeatedly doesn't do anything extra, but this still betrays expectations around when the work is being done. Calling .Next() repeatedly for example wouldn't actually do any work, which feels off. In our example, .Valid() should simply do bounds and error checking. .Start() could simply call .Next() internally, to "prime" the iterator.

Reply:
Yup, makes sense. Next is now responsible for finding the next valid (non-nil) set of inflight span recordings from the live nodes. Changed Start to internally call Next, and Valid to only do bounds + error checking.